nexus/db-queries/src/db/pagination.rs

```diff
@@ -199,6 +203,172 @@ where
+        &mut self.query
+    }
+
+    /// Finishes construction of the [Self::source] query
+    ///
+    /// Begins construction of the portion of the query which paginates over
+    /// those results.
+    pub fn with_parameters<'a, M>(
+        mut self,
+        params: &'a DataPageParams<'a, M>,
+    ) -> RawPaginatorWithParams<'a, M> {
+        self.query.sql(") ");
+        RawPaginatorWithParams { query: self.query, params }
+    }
+
+    /// A shorthand helper to paginate by UUID or Name.
+    ///
+    /// To call this function, columns with the name "id" or "name"
+    /// must be part of the SELECT query created from [Self::source].
+    ///
+    /// It's possible to achieve this same functionality manually for any
+    /// marker type using [Self::with_parameters] and
+    /// [RawPaginatorWithParams::paginate_by_column].
+    pub fn paginate_by_id_or_name(
+        self,
+        pagparams: &PaginatedBy<'_>,
+    ) -> QueryBuilder {
+        match pagparams {
+            PaginatedBy::Id(p) => {
+                self.with_parameters(p)
+                    .paginate_by_column::<diesel::sql_types::Uuid>("id")
+            }
+            PaginatedBy::Name(p) => {
+                let params =
+                    p.map_name(|n| crate::db::model::Name::ref_cast(n));
+                self.with_parameters(&params)
+                    .paginate_by_column::<diesel::sql_types::Text>("name")
+            }
+        }
+    }
+}
+
+pub struct RawPaginatorWithParams<'a, M> {
+    query: QueryBuilder,
+    params: &'a DataPageParams<'a, M>,
+}
+
+impl<M> RawPaginatorWithParams<'_, M> {
+    pub fn paginate_by_column<BindSt>(
+        mut self,
+        key_column: &'static str,
+    ) -> QueryBuilder
+    where
+        Pg: diesel::sql_types::HasSqlType<BindSt>,
+        M: diesel::serialize::ToSql<BindSt, Pg> + Send + Clone + 'static,
+        BindSt: Send + 'static,
+    {
+        if let Some(marker) = self.params.marker {
+            let compare_op = match self.params.direction {
+                dropshot::PaginationOrder::Ascending => " > ",
+                dropshot::PaginationOrder::Descending => " < ",
+            };
+
+            self.query
+                .sql("WHERE ")
+                .sql(key_column)
+                .sql(compare_op)
+                .param()
+                .bind::<BindSt, _>((*marker).clone());
+        }
+        let order = match self.params.direction {
+            dropshot::PaginationOrder::Ascending => " ASC ",
+            dropshot::PaginationOrder::Descending => " DESC ",
+        };
+        self.query
+            .sql(" ORDER BY ")
+            .sql(key_column)
+            .sql(order)
+            .sql("LIMIT ")
+            .param()
+            .bind::<diesel::sql_types::BigInt, _>(i64::from(
+                self.params.limit.get(),
+            ));
+        self.query
+    }
+}
+
+impl<M1, M2> RawPaginatorWithParams<'_, (M1, M2)> {
+    pub fn paginate_by_columns<BindSt1, BindSt2>(
```

Comment:
It seems way for feasible to make this generic across a tuple of an arbitrary length -- the generic parameters are **much** simpler.